### ISSUE TITLE
Threads Beta — Design implementation review

### ISSUE THREAD

**Issue by janogarcia** [vector-im/element-web#21502]
### Resources

- [Figma: Reference screenshots](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=4072%3A320823)

---

### Thread list

- [x] Root message avatar: Make `.mx_EventTile_avatar` render a 36px image (instead of 24px) <br> ❌ **Apr. 1**  Now it's measuring 32px instead of 36px.
- [x] Root message avatar: Change `.mx_EventTile_avatar` top position to `0` (instead of `top: -4px`)
- [x] Vertical alignment for elements in the thread summary should behave the same way as thread summaries in the main timeline (the ones that Michael (t3chguy) recently fixed).
- [ ] Align thread list to the top (latest activity on the top, [screenshot](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=4075%3A320827)). <br> ✋ Blocked by [BACAT scrolling](https://github.com/matrix-org/matrix-react-sdk/blob/develop/docs/scrolling.md#bacat-bottom-aligned-clipped-at-top-scrolling) <br> ❌ **May. 26** It's been suggested to treat this enhancement separately from the Threads project. https://github.com/vector-im/element-web/issues/21502#issuecomment-1138637261
- [x] Make the right gutter 40px ([screenshot](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=4075%3A320829): currently set to 32px). <br> ❌ **Apr. 1**  Now it's measuring 46px instead of 40px.
- [x] Unread indicator: set dot size to 10px and center it. ([screenshot](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=4075%3A320829)) <br> → [Figma](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=1935%3A254414) <br> 🐛 **Apr. 1**  Barely visible, the dot is almost completely cut off (macOS Chrome).
- [x] Unread indicator: use Compound’s secondary color for the gray color (instead of `#61708b`)

### Thread list: Hover state

- [x] ~~Fix split background colors. ([screenshot](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=4072%3A320823))~~ <br> Tracked by https://github.com/vector-im/element-web/issues/21759 
- [x] Remove left green border for own root messages.
- [x] Make the border radius 8px (instead of 4px).
- [x] ~~Extend background color on all sides to add expected padding.~~ <br> Tracked by https://github.com/vector-im/element-web/issues/21759 

### Thread list: Empty state

Refer to [mockups](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=2076%3A256634).

- [x] Fix the margins for the contents ([screenshot](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=4075%3A320826)).
- [x] Add missing line: `Tip: Use “Reply in thread” when hovering over a message.`. <br> → [Figma](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=2076%3A256634) <br> 🐛 **Apr. 1**  When visiting a room it suddenly changes to the previous version (the one without the tip) when it attempts to load the thread list for the first time.
- [x] Set `Show all threads` font size to 15px.
- [x] Filter popover menu: Do not reverse caret direction when control is active. <br> → [Figma](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=1150%3A188430)
- [x] Filter popover menu: Set a gap of 2px between popover menu and the control. <br> → [Figma](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=1150%3A188430) <br> ❌ **Apr. 1**  The gap is now bigger, get it 2px closer.

### Thread view

- [x] Make the left gutter 60px ([screenshot](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=4079%3A320835): currently set to 48px) <br> ❌ **Apr. 1**  It was 48px, now 52px, still not 60px.
- [x] The user name and message body should align to the same vertical line. ([screenshot](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=4079%3A320835))
- [x] Vertical spacing between message titles: Apply the same spacing as in the main timeline ([screenshot](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=4079%3A320833)).  <br> ❌ **Apr. 1**  Set `padding-top: 14px` for `.mx_ThreadView .mx_EventTile`.
- [x] ~~Align timeline to the top of the timeline container.~~ <br> → [Figma](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=1150%3A183119) <br> ✋ Blocked by [BACAT scrolling](https://github.com/matrix-org/matrix-react-sdk/blob/develop/docs/scrolling.md#bacat-bottom-aligned-clipped-at-top-scrolling) <br> ❌ **May. 26** We're dropping this requirement for the thread timeline. 
- [x] `View in room` action is missing in the root message contextual actions.

### Thread search: Keyword match in thread reply

- [x] Fix layout for “From a thread” footer: align left edge and add missing gap between icon and label (just change `margin-left: 8px;` for `.mx_ThreadSummaryIcon::before` to be `margin-right: 8px;`), reduce top margin (make `margin-top: 8px` for `mx_ThreadSummaryIcon`). <br> → [Figma](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=2718%3A266315) <br> ❌ **Apr. 1**  "From a thread" text should use secondary color. Also, `.mx_ThreadSummaryIcon` is now an inline element, and thus the 8px vertical margin is having no effect.
- [x] Make the thread reply footer `mx_ThreadSummaryIcon` clickable, behaving the same way as its related `mx_EventTile_line`.

### Thread search: Keyword match in thread root message

- [x] Root messages in search results should display the thread summary in the footer. (“From a thread” it’s only for thread replies) <br> → [Figma](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=2718%3A266128) <br> Tracked by https://github.com/vector-im/element-web/issues/21450

Refer to the annotations on how the flow should work for “Keyword match in root message”:

- Clicking on the root message shows it in the room timeline.
- Clicking on the thread summary opens the thread panel for the matching room. It will only exit the search result screen if keyword match is from a different room.

**Comment by t3chguy** [comment]
> Align thread list to the top ([screenshot](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=4075%3A320827)).

This is non-trivial as this is part of our ScrollPanel pagination mechanism - https://github.com/matrix-org/matrix-react-sdk/blob/develop/docs/scrolling.md#bacat-bottom-aligned-clipped-at-top-scrolling

> The approach taken instead is to vertically align the timeline tiles to the bottom of the scroll container (using flexbox) and give the timeline inside the scroll container an explicit height, initially set to a multiple of the PAGE_SIZE (400px at time of writing) as needed by the content.

**Comment by t3chguy** [comment]
> Vertical spacing between message titles: Apply the same spacing as in the main timeline ([screenshot](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=4079%3A320833)).

This already appears to be the case, we don't put any spacing between messages but have an inflated line-height to cause it.

![image](https://user-images.githubusercontent.com/2403652/159940599-4b56a69e-bae7-4253-b1b5-5123f70337c3.png)
![image](https://user-images.githubusercontent.com/2403652/159940620-cf4ace62-1808-4f08-a0b5-63acaf34226c.png)

**Comment by t3chguy** [comment]
>  Make the border radius 8px (instead of 4px).

This already appears to be 8px, can you confirm?

**Comment by nadonomy** [comment]
<img width="725" alt="Screenshot 2022-03-25 at 11 59 52" src="https://user-images.githubusercontent.com/4626865/160117056-16f3c836-48ed-49c9-9bb5-3d2be1e594aa.png">

@janogarcia @t3chguy @gsouquet can I add these visual glitches to your list pls? (develop, latest):

1. Display name split across multiple lines. I assume/suggest it should always be single line and use `text-overflow: ellipsis;` or similar.
2. At certain responsive sizes, read receipts can overlay the thread summary. I assume/suggest the thread summary should be less wide at this breakpoint.

**Comment by t3chguy** [comment]
@nadonomy 1. is https://github.com/vector-im/element-web/issues/21472

**Comment by t3chguy** [comment]
> Root messages in search results should display the thread summary in the footer. (“From a thread” it’s only for thread replies)

This is a limitation of the spec, the search API doesn't tell us they're thread roots: https://github.com/matrix-org/matrix-doc/pull/3666

**Comment by janogarcia** [comment]
> This is non-trivial as this is part of our ScrollPanel pagination mechanism - https://github.com/matrix-org/matrix-react-sdk/blob/develop/docs/scrolling.md#bacat-bottom-aligned-clipped-at-top-scrolling

@t3chguy Thanks for sharing the technical background on the current timeline scrolling implementation. 👍

That said, we shouldn’t expect all scrolling lists in the app to behave the same way, as those are dependent on the content. What makes complete sense for the room timeline may not be the most usable pattern for other contexts. I know this is from a completely different context, with its own constraints, but we’ve implemented it already for iOS/Android and need this to work consistently across platforms.

So, what if we make some tradeoffs to make it work?

→ For example, let’s say we’re able to get the desired scrolling behavior TACAB (top-aligned, clipped-at-bottom), but we don’t try to prevent the list from jumping when new elements are dynamically added/removed? Would that make it any easier? Jumps in the thread list, while not an ideal UX, would be far less frequent than jumps in a timeline which gets constantly updated.

**Comment by t3chguy** [comment]
> That said, we shouldn’t expect all scrolling lists in the app to behave the same way, as those are dependent on the content. What makes complete sense for the room timeline may not be the most usable pattern for other contexts. I know this is from a completely different context, with its own constraints, but we’ve implemented it already for iOS/Android and need this to work consistently across platforms.

Unfortunately maintaining two scrolling implementations will be very maintenance expensive, preventing scroll jumps in infinite-scrolling lists is non-trivial.

> → For example, let’s say we’re able to get the desired scrolling behavior TACAB (top-aligned, clipped-at-bottom), but we don’t try to prevent the list from jumping when new elements are dynamically added/removed? Would that make it any easier? Jumps in the thread list, while not an ideal UX, would be far less frequent than jumps in a timeline which gets constantly updated.

Yes it would but the maintenance burden would still be large, it'd cause a lot of code duplication and lead to a lot of potential for bugs due to the low test coverage we have and are able to have over complex interactions like infinite scrollers.

**Comment by janogarcia** [comment]
> This already appears to be the case, we don't put any spacing between messages but have an inflated line-height to cause it.

@t3chguy To be more specific, I mean the existing gap/padding between `.mx_EventTile` items, that _aren't_ `.mx_EventTile_continuation`.

![Google Chrome](https://user-images.githubusercontent.com/60790/160426707-4ae777ad-1b42-48eb-a7f3-dfe8a1fedd61.png)


![Threads – Figma](https://user-images.githubusercontent.com/60790/160426683-10918c5d-f993-4101-8b46-15652dd0c76e.png)

**Comment by janogarcia** [comment]
> This already appears to be 8px, can you confirm?

@t3chguy Let's say `.mx_EventTile[data-shape=ThreadsList]` has 8px applied to it, but it's not being rendered as intended/expected.

In the following example, to make it more obvious, I have set `border-radius: 9999px` to `.mx_EventTile[data-shape=ThreadsList]` and have changed the background colors of the split background issue. As you can see there's an awful mix of border radii values being applied over there.

![Element  1  | 🧪 Playground Encrypted](https://user-images.githubusercontent.com/60790/160429600-b1ffe88d-ec28-4db3-a980-2cc6512b16e3.png)

**Comment by t3chguy** [comment]
> @t3chguy In that case, we can wait for the spec to be updated.


Tracked by https://github.com/vector-im/element-web/issues/21450

**Comment by t3chguy** [comment]
![image](https://user-images.githubusercontent.com/2403652/160621068-22ea508d-bb39-4b21-865b-0f2f77826668.png)

Editing in threads doesn't line up with Save button and the grey bg doesn't fill the width

**Comment by t3chguy** [comment]
> In the following example, to make it more obvious, I have set border-radius: 9999px to .mx_EventTile[data-shape=ThreadsList] and have changed the background colors of the split background issue. As you can see there's an awful mix of border radii values being applied over there.

Unfortunately the top and bottom radii must be on two different selectors as we can't use `overflow: hidden` due to the action bar floating out of the parent container. Have used a postcss variable to set both to 8px now.

**Comment by janogarcia** [comment]
@t3chguy  Yeah, I was just answering your `This already appears to be 8px, can you confirm?` question by illustrating that while having 8px applied to it, it didn’t necessarily mean the resulting box would have 8px rounded corners.

It actually resulted in 3 different values (top corners for yellow background, right corners and left corners for orange background).

**Comment by janogarcia** [comment]
> Display name split across multiple lines. I assume/suggest it should always be single line and use text-overflow: ellipsis; or similar.

@nadonomy Thanks for flagging, Nad. 👍

@t3chguy So, I've been experimenting a bit with the truncating behavior when there are long display names. I basically wanted to test two approaches, prioritizing display name over reply body vs. equal priority for display name and reply body. 

After mocking it up on Figma and testing it live on the web app, I'd recommend going for the former (prioritizing display name over reply body) because:

a) we're hopefully truncating one piece of content instead of two, which would be harder to read
b) it shouldn't be that common

Neither approach is perfect, having their own pros/cons, but we can easily switch between both implementations (it's just a matter of toggling a CSS declaration).

The proposed behavior can be implemented by adding these rulesets:

```
.mx_ThreadInfo_sender {
    text-overflow: ellipsis;
    overflow: hidden;
    white-space: nowrap;
}

.mx_ThreadInfo_content {
    flex: 1;
}
```

The alternate approach would be just a matter of removing the `flex: 1` declaration.

--- 

**Web App**

![Element  1  | Threads P0 Delivery](https://user-images.githubusercontent.com/60790/160653479-1911df27-2287-4091-b2a7-6005c62946bc.png)

![Element  1  | Threads P0 Delivery](https://user-images.githubusercontent.com/60790/160653553-d6f7646d-4862-4cca-958b-5600131d19ff.png)

---

**Figma**

![Threads – Figma](https://user-images.githubusercontent.com/60790/160646407-a0dd028c-f792-4e2c-b0a2-77d3c19e2332.png)

**Comment by janogarcia** [comment]
Actually, let's take the opposite approach (equal priority). Reason being that:

- users gets more useful context in the summary
- worst case scenario looks worse when prioritizing the name (it can take up all the available space)

I'm editing my previous reply.

**Comment by janogarcia** [comment]
> At certain responsive sizes, read receipts can overlay the thread summary. I assume/suggest the thread summary should be less wide at this breakpoint.

@t3chguy That shouldn't be happening. It seems `.mx_ThreadInfo` is not tracking/following the same width constraints as `.mx_EventTile_line`. Can we please make it track that constraint? 

I think @jryans implemented the current responsive optimizations for the thread summary, maybe he can add some context for that difference, if intended for some reason.

![Google Chrome](https://user-images.githubusercontent.com/60790/160659251-20b49ae4-90c1-43af-8243-42cbe0d34a22.png)

**Comment by jryans** [comment]
> Thread view: The user name and message body should align to the same vertical line. ([screenshot](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=4079%3A320835))

@t3chguy: The above seems marked as fixed, but it's still happening here...?

<img width="252" alt="image" src="https://user-images.githubusercontent.com/279572/160799044-50b915bf-12f2-4de1-ae05-0ba12418a5a0.png">

**Comment by janogarcia** [comment]
@jryans I think the related PR hasn’t been merged yet.

**Comment by t3chguy** [comment]
@jryans its fixed in the PR only

**Comment by jryans** [comment]
> @t3chguy That shouldn't be happening. It seems `.mx_ThreadInfo` is not tracking/following the same width constraints as `.mx_EventTile_line`. Can we please make it track that constraint?
> 
> I think @jryans implemented the current responsive optimizations for the thread summary, maybe he can add some context for that difference, if intended for some reason.

Hmm no, it's not intentional... Both `.mx_EventTile_line` and `.mx_ThreadInfo` have a `margin-right` of 110px to allow room for the read receipts... but `.mx_ThreadInfo` is poking out of its parent box for some reason, so the margin isn't limiting the width properly... I'll see if I can spot a fix for this.

**Comment by jryans** [comment]
> @jryans I think the related PR hasn’t been merged yet.
> @jryans its fixed in the PR only

Ah okay, I did look for a PR first... but may have only looked in the review requests list, whereas it's [this draft PR](https://github.com/matrix-org/matrix-react-sdk/pull/8149). 😅

**Comment by t3chguy** [comment]
>  Vertical alignment for elements in the thread summary should behave the same way as thread summaries in the main timeline (the ones that Michael (t3chguy) recently fixed).

@janogarcia can you elaborate on this?

It is the only non-blocked point which is not tracked in another issue

**Comment by janogarcia** [comment]
@t3chguy Sure, we need to make sure we're applying the same vertical alignment properties/behavior for thread summaries in the thread list as those we defined a while ago for thread summaries in the main timeline. Ideally, we wouldn't need to define/update these styles in two separate places.

- Left (main timeline): correct alignment, that we fixed a while ago.
- Right (thread list): reply count and message preview are not sitting on the expected baseline (just the same issue we had for thread summaries in the main timeline).

![SCR-20220425-nax-2](https://user-images.githubusercontent.com/60790/165116449-0e8d3e41-00ce-4548-85de-3bfaf5ec7517.png)

**Comment by novocaine** [comment]
@janogarcia can you please confirm whether we have done enough here to close the ticket?

**Comment by janogarcia** [comment]
@novocaine @t3chguy 

So, the only issue pending to be addressed is the infinite scrolling behavior. More specifically, the limitation of the current implementation ([BACAT](https://github.com/matrix-org/matrix-react-sdk/blob/develop/docs/scrolling.md#bacat-bottom-aligned-clipped-at-top-scrolling)), that prevents us from ordering the scrollable events from top to bottom. 

I’ve been thinking more about this lately and would like to share some updates:

- I’m dropping the requirement (I already checked it off) for the top-to-bottom alignment in the thread timeline (`Align timeline to the top of the timeline container.`), which is, by far, the less critical one. I’d even say that it can work almost equally well either with BACAT or TACAB scrolling.
- I still consider the case for the thread list to be relevant, though (`Align timeline to the top of the timeline container.`). We (the design team) would still want users to have scrolling lists where the items are ordered from top to bottom. That doesn’t only make sense for the Thread List but other contexts as well, like the Mentions/Notifications panel or the Files panel.

From a UX perspective, we can’t presume that BACAT scrolling will work and serve best every possible scenario and interaction need (e.g., we already use the top-to-bottom scrolling pattern in the room list on the left sidebar and the people list on the right sidebar, although not for infinite items).

We consider top-to-bottom scrolling a core interaction pattern worth the effort of solving. It surely has a considerable implementation cost, but it also has a similar or higher product/UX cost not having it.

Now, given that we want to reuse this behavior in different contexts (mentions panel, files panel), meaning it’s not a Threads-specific limitation but a core interaction issue affecting our product in general, I think we could move this requirement out of the scope of the Exit Beta milestone for Threads and handle it separately.

So, to recap:

- We are dropping the requirement for the vertical alignment of the **thread timeline** altogether.
- We’d still want to order the **thread list** from most recently updated on the top to oldest on the bottom. We would handle it separately from the Threads project, treating it as a global app enhancement instead.

What do you think?

---

/cc @nadonomy CCing you as per a previous internal discussion.

**Comment by t3chguy** [comment]
> I still consider the case for the thread list to be relevant, though (Align timeline to the top of the timeline container.). We (the design team) would still want users to have scrolling lists where the items are ordered from top to bottom. That doesn’t only make sense for the Thread List but other contexts as well, like the Mentions/Notifications panel or the Files panel.


The ordering isn't an issue to flip, the issue is the potential for a gap at the top due to how BACAT reserves space to prevent scroll jumps e.g by the spinner coming up and going away.

TACAB scrolling has guaranteed scroll-jumps for lazily/infinite loaded lists as the spinner renders and then vanishes.

> We’d still want to order the thread list from most recently updated on the top to oldest on the bottom. We would handle it separately from the Threads project, treating it as a global app enhancement instead.

Great, though would be good to clarify whether by 

> We (the design team) would still want users to have scrolling lists where the items are ordered from top to bottom.

You mean just flipping the order or TACAB which would trade scroll jumps for having no gap

Thanks

**Comment by janogarcia** [comment]
We want to be able to flip the order, top to bottom, with the first item being aligned to the top of the scrolling container, without any extraneous/additional white space on the top, as in the mockup below (a filterable/sortable list of items).

I'd also expect no scroll position jumps when interacting with the scrolling behavior. If the spinner is introducing the problem, then we could think of a slightly different design for it.

---

<img width="400" alt="Frame 6598" src="https://user-images.githubusercontent.com/60790/170515217-08df5dea-8a68-441b-8ea4-e84bb4f14de2.png">

**Comment by t3chguy** [comment]
If we can replace the spinner with something that consumes no vertical space (to have to pop-in and out) then that would help

**Comment by t3chguy** [comment]
We can probably hack BACAT to remove the paging behaviour for the threads list, as long as we can turn the spinner/no-spinner into a constant height element. Like the old youtube top-bar horizontal animation which I cannot find a reference to atm

**Comment by janogarcia** [comment]
> If we ca n replace the spinner with something that consumes no vertical space (to have to pop-in and out) then that would help

What if we did something like this?

Let's say `Item 09` was already available on first load. Then, we'd draw the spinner over the existing content when fetching more items (`Item 10` and up). Would that make it any easier? 

<img width="400" alt="Frame 6598" src="https://user-images.githubusercontent.com/60790/170545165-55035f94-0686-4718-9422-95da9ec794dc.png">

**Comment by janogarcia** [comment]
> Like the old youtube top-bar horizontal animation which I cannot find a reference to atm

Can't recall that one, but would to see an example.

**Comment by t3chguy** [comment]
Yup an overlaid loading state would work too, anything that doesn't cause the vertical bounce

**Comment by janogarcia** [comment]
Alternatively, we could just show an animated loading bar on the bottom that would only take up 2 additional pixels. We could just keep that 2px gap on the bottom even when the loader is inactive. We could live with that, even if not ideal.

<img width="400" alt="Frame 6598 (2)" src="https://user-images.githubusercontent.com/60790/170547161-4a1ff7ac-d490-497d-904c-719608aa8d2b.png">

We could also combine both for better feedback on what's being loaded, but it's probably getting a bit noisy.

<img width="400" alt="Frame 6598 (1)" src="https://user-images.githubusercontent.com/60790/170547558-dbd16fd8-16c9-4f60-8f8b-0afc61345e9b.png">

**Comment by t3chguy** [comment]
https://www.cssscript.com/youtube-top-loader/ the loading bar was what I had in mind, combining both is also possible

**Comment by janogarcia** [comment]
> https://www.cssscript.com/demo/youtube-top-loader/ 

Thanks for the example! Yeah, basically the same stuff I suggested in the previous example. 👍

**Comment by t3chguy** [comment]
I just noticed Github has such a minimal loading animation also :D

**Comment by daniellekirkwood** [comment]
believe this is completed and can therefore be closed...

**Comment by janogarcia** [comment]
@daniellekirkwood Related [internal discussion on BACAT scrolling and sorting](https://matrix.to/#/!tCaHWuHHvOHjWyqapP:matrix.org/$sddU6zNKb3pJpPdJ_B12qe8GcwnVi9PTY_iZqYKFxsM). 

That said, we probably don't plan to act on the scrolling implementation soon, so we may want to consider that as a standalone enhancement (affecting the threads list, but also the mentions and files panels) that we can separate from the Threads project, as [suggested previously](https://github.com/vector-im/element-web/issues/21502#issuecomment-1138637261).

**Comment by daniellekirkwood** [comment]
Created an issue to keep track of the scrolling things [here](https://github.com/vector-im/element-meta/issues/934)

Will now close this issue.
